### PR TITLE
osu! commands [rewritten!]

### PR DESCRIPTION
**Pull request by jaxxibae**
+ Rewritten to work with new bot features and v12
+ Upgraded from #54 and #801 
+ To be merged with SwitchbladeBot/bladey-web#71

# osu! user lookup (taiko, ctb, mania and osu!)

image
image
image

# osu! beatmap lookup (same modes)

image
image
image
image

### CODE REVIEW

**Comment by jaxxibae** [comment]
I actually need to take new screenshots, since those are with the old embed.

File: src/apis/Osu.js [hunk 1]
```diff
@@ -0,0 +1,74 @@
...
+  }
+
+  getUser (user, mode) {
+    return this.request('/get_user', { u: user, m: mode }).then(u => u[0])
+  }
+
+  async getBeatmap (beatmap, mode, limit = 5) {
+    return this.request('/get_beatmaps', { b: beatmap, m: mode, limit }).then(data => {
+      if (typeof data !== 'undefined' && data.length > 0) return data
+      else return this.getBeatmapSet(beatmap, limit)
```
Comment (fernando457829) [inline src/apis/Osu.js:24]:
```suggestion
      else return this.getBeatmapSet(beatmap, null, limit)
```

File: src/utils/EndpointUtils.js [hunk 2]
```diff
@@ -21,6 +21,8 @@ module.exports = class EndpointUtils {
                 if (fetchGuilds) req.guilds = await this._fetchGuilds(client, accessToken)
                 return next()
               } catch (e) {
+                console.log(e)
```
Comment (fernando457829) [inline src/utils/EndpointUtils.js:24]: Remove this `console.log`

File: src/commands/games/osu/player.js [hunk 3]
```diff
@@ -0,0 +1,118 @@
...
+            [
+              description.join('\n')
+            ]
+          ]))
+      }
+
+      paginatedEmbed.run(await channel.send('...'))
+      channel.stopTyping()
+    } catch (e) {
+      console.log(e)
```
Comment (fernando457829) [inline src/commands/games/osu/player.js:113]: Remove this `console.log`

File: src/commands/games/osu/player.js [hunk 4]
```diff
@@ -0,0 +1,118 @@
...
+            ]
+          ]))
+      }
+
+      const recentPlays = await this.client.apis.osu.getUserRecentPlays(user, mode[0], 5)
+
+      if (recentPlays.length > 0) {
+        let description = []
+
+        for (var int in recentPlays) {
```
Comment (fernando457829) [inline src/commands/games/osu/player.js:91]: prefer `const`
```suggestion
        for (const int in recentPlays) {
```

File: src/commands/games/osu/player.js [hunk 5]
```diff
@@ -0,0 +1,118 @@
...
+            t('commands:osu.subcommands.player.playCount', { playCount: MiscUtils.formatNumber(userData.playcount, language) }),
+            t('commands:osu.subcommands.player.totalHits', { totalHits: MiscUtils.formatNumber((parseInt(userData.count300) + parseInt(userData.count100) + parseInt(userData.count50)), language), 300: MiscUtils.formatNumber(userData.count300, language), 100: MiscUtils.formatNumber(userData.count100, language), 50: MiscUtils.formatNumber(userData.count50, language), emoji300: this.client.officialEmojis.get('300'), emoji100: this.client.officialEmojis.get('100'), emoji50: this.client.officialEmojis.get('50') })
+          ]
+        ]))
+
+      const topScores = await this.client.apis.osu.getUserTopScores(user, mode[0], 5)
+      if (topScores.length > 0) {
+        let description = []
+
+        for (var i in topScores) {
```
Comment (fernando457829) [inline src/commands/games/osu/player.js:68]: prefer `const`
```suggestion
        for (const i in topScores) {
```

File: src/commands/games/osu/player.js [hunk 6]
```diff
@@ -0,0 +1,118 @@
...
+            [
+              description.join('\n')
+            ]
+          ]))
+      }
+
+      const recentPlays = await this.client.apis.osu.getUserRecentPlays(user, mode[0], 5)
+
+      if (recentPlays.length > 0) {
+        let description = []
```
Comment (fernando457829) [inline src/commands/games/osu/player.js:89]: prefer `const`
```suggestion
        const description = []
```

File: src/commands/games/osu/player.js [hunk 7]
```diff
@@ -0,0 +1,118 @@
...
+            t('commands:osu.subcommands.player.totalHits', { totalHits: MiscUtils.formatNumber((parseInt(userData.count300) + parseInt(userData.count100) + parseInt(userData.count50)), language), 300: MiscUtils.formatNumber(userData.count300, language), 100: MiscUtils.formatNumber(userData.count100, language), 50: MiscUtils.formatNumber(userData.count50, language), emoji300: this.client.officialEmojis.get('300'), emoji100: this.client.officialEmojis.get('100'), emoji50: this.client.officialEmojis.get('50') })
+          ]
+        ]))
+
+      const topScores = await this.client.apis.osu.getUserTopScores(user, mode[0], 5)
+      if (topScores.length > 0) {
+        let description = []
+
+        for (var i in topScores) {
+          const beatmap = await this.client.apis.osu.getBeatmap(topScores[i].beatmap_id, 0)
```
Comment (fernando457829) [inline src/commands/games/osu/player.js:69]: Maybe get first element on `beatmap` instead `beatmap[0]`

```js
const [firstElement] = await getBeatmap()

if (firstElement) description.push('...')
```

File: src/commands/games/osu/player.js [hunk 8]
```diff
@@ -0,0 +1,118 @@
...
+          ]))
+      }
+
+      const recentPlays = await this.client.apis.osu.getUserRecentPlays(user, mode[0], 5)
+
+      if (recentPlays.length > 0) {
+        let description = []
+
+        for (var int in recentPlays) {
+          const beatmap = await this.client.apis.osu.getBeatmap(recentPlays[int].beatmap_id, mode[0])
```
Comment (fernando457829) [inline src/commands/games/osu/player.js:92]: Maybe get first element on beatmap instead beatmap[0] (again haha)

```js
const [firstElement] = await getBeatmap()

if (firstElement) description.push('...')
```

**Comment by jaxxibae** [comment]
# RFR

File: src/apis/Osu.js [hunk 9]
```diff
@@ -0,0 +1,78 @@
...
+const qs = require('querystring')
+const { URLSearchParams } = require('url')
+
+const API_URL = 'https://osu.ppy.sh/api'
+
+module.exports = class Osu extends APIWrapper {
+  constructor () {
+    super({
+      name: 'osu',
+      envVars: ['OSU_API_KEY']
```
Comment (pedrofracassi) [inline src/apis/Osu.js:13]:
```suggestion
      envVars: [ 'OSU_API_KEY', 'OSU_CLIENT_ID' ]
```

File: src/locales/en-US/commands.json [hunk 10]
```diff
@@ -362,6 +362,44 @@
...
+        "playerNotFound": "Profile not found."
+      },
+      "beatmap": {
+        "commandDescription": "Searches for a beatmap or beatmap set on osu!",
+        "commandUsage": "<beatmap or beatmap set id>",
+        "noBeatmap": "You have to give me a valid beatmap or beatmap set ID.",
+        "starDifficulty": "Star Difficulty: **{{difficulty}}**",
+        "successRate": "Success Rate: **{{rate}}%** ({{successfulPlays}} / {{totalPlays}})",
+        "mappedBy": "Mapped by {{mapper}}",
+        "lastUpdate": "Last update on **{{date}}** ({{timeAgo}})",
```
Comment (pedrofracassi) [inline src/locales/en-US/commands.json:394]:
```suggestion
        "lastUpdate": "Last update at **{{date}}** ({{timeAgo}})",
```

File: src/locales/en-US/commands.json [hunk 11]
```diff
@@ -362,6 +362,44 @@
...
+        "approved": "Approved on **{{date}}** ({{timeAgo}})",
```
Comment (pedrofracassi) [inline src/locales/en-US/commands.json:395]:
```suggestion
        "approved": "Approved at **{{date}}** ({{timeAgo}})",
```

File: src/locales/en-US/commands.json [hunk 12]
```diff
@@ -362,6 +362,44 @@
...
+  "osu": {
+    "commandDescription": "Searches for a profile or a beatmap on osu!",
+    "commandUsage": "<type> <query>",
+    "subcommands": {
+      "player": {
+        "commandDescription": "Searches for a profile on osu!",
+        "commandUsage": "[profile name or id]",
+        "noPlayer": "You have to give me a valid osu! profile name or ID.",
+        "level": "Level **{{number}}**",
+        "joinedAt": "Joined on **{{date}}** ({{timeAgo}})",
```
Comment (pedrofracassi) [inline src/locales/en-US/commands.json:374]:
```suggestion
        "joinedAt": "Joined at **{{date}}** ({{timeAgo}})",
```

File: src/locales/en-US/commands.json [hunk 13]
```diff
@@ -362,6 +362,44 @@
...
+        "commandDescription": "Searches for a beatmap or beatmap set on osu!",
+        "commandUsage": "<beatmap or beatmap set id>",
+        "noBeatmap": "You have to give me a valid beatmap or beatmap set ID.",
+        "starDifficulty": "Star Difficulty: **{{difficulty}}**",
+        "successRate": "Success Rate: **{{rate}}%** ({{successfulPlays}} / {{totalPlays}})",
+        "mappedBy": "Mapped by {{mapper}}",
+        "lastUpdate": "Last update on **{{date}}** ({{timeAgo}})",
+        "approved": "Approved on **{{date}}** ({{timeAgo}})",
+        "tags": "Tags: **{{tags}}**",
+        "download": "[Download]({{link}}) - [no video]({{linkNoVideo}})",
```
Comment (pedrofracassi) [inline src/locales/en-US/commands.json:397]:
```suggestion
        "download": "[Download]({{link}}) - [Download without video]({{linkNoVideo}})",
```

Comment (pedrofracassi) [reply src/locales/en-US/commands.json:397]: What happens if a beatmap already has no video?

Comment (jaxxibae) [reply src/locales/en-US/commands.json:397]: It'll download just fine, but I can make that check if it has or hasn't got a video

Comment (pedrofracassi) [reply src/locales/en-US/commands.json:397]: Please do :)

File: src/apis/Osu.js [hunk 14]
```diff
@@ -0,0 +1,78 @@
...
+const qs = require('querystring')
+const { URLSearchParams } = require('url')
+
+const API_URL = 'https://osu.ppy.sh/api'
+
+module.exports = class Osu extends APIWrapper {
+  constructor () {
+    super({
+      name: 'osu',
+      envVars: ['OSU_API_KEY', 'OSU_CLIENT_ID', 'OSU_CLIENT_SECRET']
```
Comment (Doges) [inline src/apis/Osu.js:13]: cade no .env.example

Comment (jaxxibae) [reply src/apis/Osu.js:13]: You're a pain in the ass

**Review by Doges — APPROVED** [review]
[REDACTED]

File: src/utils/EndpointUtils.js (hunk 2 above)
Comment (pedrofracassi) [reply src/utils/EndpointUtils.js]: @jaxxibae

**Pull request merged**